CONTINUATION.md

```diff
@@ -0,0 +1,511 @@
+
+## Promise operations
+
+Promise is the basic control-flow building block in async codes. Since promises
+are first-class values, they can be passed around, aggregated, and so on.
+Promise handlers can be attached to a promise in a different context than the
+creation context of the promise.
+
+The following promise aggregation APIs may be a merge point where multiple
+promises are merged into one single promise, or short-circuit in conditions
+picking one single winner and discarding all other states.
+
+| Name                 | Description                                     | On-resolve    | On-reject     |
+| -------------------- | ----------------------------------------------- | ------------- | ------------- |
+| `Promise.allSettled` | does not short-circuit                          | Merge         | Merge         |
+| `Promise.all`        | short-circuits when an input value is rejected  | Merge         | Short-circuit |
+| `Promise.race`       | short-circuits when an input value is settled   | Short-circuit | Short-circuit |
+| `Promise.any`        | short-circuits when an input value is fulfilled | Short-circuit | Merge         |
+
+To expand on the behaviors, given a task with the following definition, with
+`ctx` being the corresponding context variable (`AsyncContext.Variable` and
+`ContinuationFlow`) in each semantic:
+
+```js
+let taskId = 0
+function randomTask() {
+  return ctx.run(`task-${taskId++}`, async () => {
+    await scheduler.wait(Math.random * 10)
+    if (Math.random() >= 0.5) {
+      throw new Error()
+    }
+  })
+}
+```
+
+### `Promise.allSettled`
+
+It is a merge point on `Promise.allSettled`. It aggregates the promise results
+that are either resolved or rejected.
+
+```js
+ctx.run('main', async () => {
+  await Promise.allSettled(_.times(5).map((it) => randomTask()))
+  ctx.get() // => (1)
+})
+```
+
+From different observation perspectives, the value at site (1) would be
+different on the above semantics:
+
+- For `AsyncContext.Variable`, it is `main`,
+- For `ContinuationFlow`, the proposed solution didn't specify
+  conflicts-resolving yet. If the default behavior merges the values in an
+  array, it changes the shape of the value.
+
+### `Promise.all`
+
+In the resolving path, `Promise.all` is similar to `Promise.allSettled` that it
+is a merge point when all promises resolve.
+
+`Promise.all` picks the fastest reject promise.
+
+```js
+ctx.run('main', async () => {
+  try {
+    await Promise.all(_.times(5).map((it) => randomTask()))
+    ctx.get() // => (1)
+  } catch (e) {
+    ctx.get() // => (2)
+  }
+})
+```
+
+The value at site (1):
+
+- For `AsyncContext.Variable`, it is `main`,
+- For `ContinuationFlow`, the proposed solution didn't specify
+  conflicts-resolving yet.
+
+The value at site (2):
+
+- For `AsyncContext.Variable`, it is `main`,
+- For `ContinuationFlow`, it is the one caused the rejection, and discarding
+  leaf contexts of promises that may have been fulfilled.
```

Comment:
As you described, the "flow" is reaching the point of rejection and then aborting the aggregate, so that would be the context I would expect. But layered promises would produce something similar to stack unwinding where we would lose context of the _exact_ source, which is why I feel an `error.originatingContext` is a more _useful_ source for the particular purpose of error analysis.